File: binary_search_algo/kotlin_bin_search/recursive_method.kt
fun main() {
    val array = arrayOf(1,3,4,5,6,7,8)
    val target = 8
    val high = array.size - 1
    val low = 0
    val result:Int = rBinarySearch(target, array, low, high)
    if(result == -1){
        println("The element is not in the array")
    }else{
        println("The element is at index: $result")
    }

}

fun rBinarySearch(target: Int, array: Array<Int>, low: Int, high: Int): Int {

    if(low <= high) {
        val sum = high + low
        val middle:Int =  sum / 2
        if(target == array[middle]) return middle
        return if(target < array[middle]) rBinarySearch(target, array, low, middle-1)
        else rBinarySearch(target, array, middle + 1, high)
    }
    return -1
}

File: binary_search_algo/cpp_bin_search/recursive_method.cpp
#include <stdio.h>
int binarySearch(int target, int* array, int high, int low);
int main(void) {

    int array[] = {1,2,3,4,5};
    int low = 0;
    int high = sizeof(array) / sizeof(*array);
    int result = binarySearch(5, array, high, low);

    printf("The target is at: %i\n", result);

}

int binarySearch(int target, int* array, int high, int low) {

    int middle = low + (high - low) / 2;
    if(low < high) {
        if(target == array[middle]) return middle;
        if(array[middle] > target) binarySearch(target, array, middle-1, low);
        else binarySearch(target, array, high, middle+1);
    }

    return -1;
}
